notebooks/ch11/gradient_stochastic.ipynb

```diff
@@ -0,0 +1,285 @@
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "\n",
+    "This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Stochastic Gradient Descent\n",
+    "\n",
+    "To circumvent the difficulty of computing a gradient across the entire training set, **stochastic gradient descent** approximates the overall gradient using a single randomly chosen data point. Since the observation is chosen randomly, we expect that using the gradient at each individual observation will eventually converge the algorithm to the same parameters as batch gradient descent. The general update formula for stochastic gradient descent is below, where $l(\\hat\\theta, y_i)$ is the loss function for a single data point:\n",
+    "\n",
+    "$$\n",
+    "\\hat\\theta_{t+1} = \\hat\\theta_t - \\alpha \\nabla_\\hat\\theta l(\\hat\\theta, y_i)\n",
+    "$$\n",
+    "\n",
+    "Returning back to our example using the MSE, we approximate the gradient of the mean squared error using the gradient of the squared loss of one data point. \n",
+    "\n",
+    "$$\n",
+    "\\begin{align}\n",
+    "\\nabla_{\\hat\\theta}L(\\hat\\theta, y) &\\approx \\nabla_{\\hat\\theta} l(\\hat\\theta, y_i) \\\\\n",
+    "&= -2(y_i - \\hat\\theta)\n",
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded, since the randomness means that $E[\\nabla_{\\hat\\theta}l(\\hat\\theta, y_i)] = \\nabla_{\\hat\\theta}L(\\hat\\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through the shuffled data is called an **epoch**; at the end of every epoch, we re-shuffle the data to ensure that the model sees examples from various classes in an arbitrary order."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Visualizing Stochastic Gradient Descent\n",
+    "\n",
+    "Below are visual examples of loss minimization in batch gradient descent and stochastic gradient descent.\n",
+    "\n",
+    "![](gd.png)\n",
+    "![](sgd.png)\n",
+    "\n",
+    "At each iteration of batch gradient descent, we move in the direction of the true gradient of the cost function, which is depicted by the ellipses. On the other hand, each iteration of stochastic gradient descent may not lead us in the direction of the true gradient; however, the $\\hat\\theta$ parameters eventually converge to the minima of the cost function. Surprisingly, stochastic gradient descent often converges more quickly than batch gradient descent in practice since stochastic gradient descent spends significantly less time updating model parameters."
```

Comment:
I remember a vitamin question that said "Takes less update steps to converge" is not a strength of SGD, so if you are referring to SGD computation time it would be good to clarify that this isn't necessarily due to fewer update steps.